### ISSUE TITLE
Draw fps capped at 60 with Direct3D 11 even though the limiter is not set to vsync

### ISSUE THREAD

**Issue by mouzedrift** [ppy/osu#22962]
### Type

Performance

### Bug description

When I use Direct3D 11, my fps are capped at 60 fps even when my fps limiter is not set to vsync. This only happens in fullscreen and borderless fullscreen, and the windowed mode doesn't seem to be limiting my fps. The OpenGL and OpenGL (Legacy) renderers don't seem to do that. I can't test Vulkan because my GPU doesn't support it.

### Version

2023.326.0

### Logs

[performance.log](https://github.com/ppy/osu/files/11072364/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11072365/runtime.log)
[updater.log](https://github.com/ppy/osu/files/11072366/updater.log)
[database.log](https://github.com/ppy/osu/files/11072367/database.log)
[input.log](https://github.com/ppy/osu/files/11072368/input.log)
[network.log](https://github.com/ppy/osu/files/11072369/network.log)

**Comment by tuhanayim** [comment]
Can reproduce, but it happens on windowed mode too for me.

**Comment by Hasster1** [comment]
Try disabling fullscreen optimizations.

**Comment by tuhanayim** [comment]
> Try disabling fullscreen optimizations.

Still can reproduce.

**Comment by peppy** [comment]
Please provide screenshots with frame graphs (ctrl+f11 twice) and also provide more details on whether you feel high input latency or are just looking at the fps counter and noticing.

**Comment by akippnn** [comment]
I also have a similar issue. (Sorry, forgot to take frame graphs for the earlier screnshots, but they were stable for the most part.)

Earlier, the the framerate seemed to be locked at 72FPS no matter what renderer I set it on, except for Vulkan, however, osu!lazer freezes on Vulkan the moment I press Play (same issue as #22960).
![image](https://user-images.githubusercontent.com/80670461/227794990-db6c746f-e5e8-4a57-b01c-41999666ea5b.png)

This was regardless of the game being in fullscreen mode. So I messed around to see what was causing it.

Somehow enabling G-SYNC changed how the game was behaving ever since. I set that on and now my game ran at the set in-game refresh rate regardless of whether it was in full-screen or not.
![image](https://user-images.githubusercontent.com/80670461/227795408-c938f3db-bd66-43e6-9cbe-1d0a281ac77e.png)
![image](https://user-images.githubusercontent.com/80670461/227795420-cb64de4f-389e-4a64-9349-c489403e7f87.png)

But now I set the game to the global default and restarted osu again, to find that the issue was somehow fixed but only in exclusive fullscreen though. Otherwise the game would be stuck in 144FPS until then. Here's the frame graphs for the latter case:
![image](https://user-images.githubusercontent.com/80670461/227796469-a2ba0885-ce80-4f60-bf71-3378ca1aa9c7.png)

Logs (nothing interesting in particular from what I've seen):
[updater.log](https://github.com/ppy/osu/files/11072646/updater.log)
[database.log](https://github.com/ppy/osu/files/11072647/database.log)
[input.log](https://github.com/ppy/osu/files/11072648/input.log)
[performance.log](https://github.com/ppy/osu/files/11072649/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11072650/runtime.log)

**Comment by longnguyen2004** [comment]
I'd say this is very much driver dependent. Try resetting all the settings and try again, who knows what those drivers are doing below osu itself...

**Comment by akippnn** [comment]
> I'd say this is very much driver dependent. Try resetting all the settings and try again, who knows what those drivers are doing below osu itself...

Yeah, that's the first conclusion I came to, but it didn't really fix much since the game was still stuck in 72FPS regardless (even after a computer restart) before I started messing around with G-SYNC settings.

**Comment by mouzedrift** [comment]
> Please provide screenshots with frame graphs (ctrl+f11 twice) and also provide more details on whether you feel high input latency or are just looking at the fps counter and noticing.

| Fullscreen | Borderless fullscreen | Windowed |
| ----------- | ----------------------- | ------------ |
| ![osu!_AaWlr7kotm](https://user-images.githubusercontent.com/43358824/227797898-ddfbef02-ed4e-42e0-826e-906b3a99bbcb.png) | ![image](https://user-images.githubusercontent.com/43358824/227797963-0c0c6d87-99bc-482b-bf52-7074bc0d8740.png) | ![image](https://user-images.githubusercontent.com/43358824/227798023-1d5b32d6-27f7-45a8-8c06-8e47b61c17f6.png) |

The mouse input latency in fullscreen is pretty bad and very delayed (borderless and windowed feel more responsive than fullscreen), but the keyboard input latency seems fine.

**Comment by longnguyen2004** [comment]
For Borderless and Windowed: You might be seeing DWM VSync in action, which can be disabled by passing [`DXGI_PRESENT_ALLOW_TEARING`](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/dxgi-present#:~:text=in%20Windows%C2%A08.1.-,DXGI_PRESENT_ALLOW_TEARING,-0x00000200UL) to the DXGI `Present` function.

I saw in the framework's code that there's an [`AllowTearing`](https://github.com/ppy/osu-framework/blob/745fce6716cdcd3c945231d7fd616c4a8e6b6168/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs#L38-L42) flag, but [it's only set when the game is in full screen mode](https://github.com/ppy/osu-framework/blob/745fce6716cdcd3c945231d7fd616c4a8e6b6168/osu.Framework/Platform/GameHost.cs#L484). You can use something like Special K to inject the flag (only a temporary solution, proper fix needs to be done in code). With that, VSync is properly off in borderless mode.

![image](https://user-images.githubusercontent.com/15729831/227828115-7ddb555f-d838-4517-9f94-7c24fa58e4bb.png)

As for Fullscreen mode, that's definitely some driver stuff going on, because I can't reproduce it, so I highly doubt osu! can do anything here.

**Comment by peppy** [comment]
We *intentionally* don't want to allow tearing in borderless, as that's generally the expectation of what "borderless" means to a user.

**Comment by peppy** [comment]
That's not the direction we will be taking, but thanks for your thoughts.

Avoiding tearing is paramount for a normal user experience.

**Comment by Andreyf20** [comment]
Same problem, tried to play around with gsync and got it to work (unlimited) but only in full screen decided to go with opengl for now which seems to be working in windowed for me.

**Comment by longnguyen2004** [comment]
Seems like people are using fullscreen windowed not to avoid tearing, but just to be able to Alt+Tab without the screen flickering. I think we can safely assume that the people using unlimited FPS already know the drawbacks and consequences of it, so we can probably allow for tearing in fullscreen windowed, enable VSync by default, and give a big fat warning if people switch away from it.

Remember, if there's no tearing, then VSync must be somewhere in the chain. It might be game VSync, DWM VSync, driver VSync, OpenGL triple buffering, etc etc., but it's there, even if the in-game counter tells you otherwise. so I'd rather make it explicit that VSync is enabled, rather than having DWM silently enabling it.

**Comment by Layendan** [comment]
I can recreate this bug with the latest version of osu.

However, if I switch framerate from 2x to vsync, then back to 2x, the issue is solved for the session/user restarts game.
Although better, I still feel as if there is input lag on both keyboard and mouse when using this fix.

Can anyone recreate this or am I the only one where changing framerates fixes the problem?

**Comment by longnguyen2004** [comment]
> Can anyone recreate this or am I the only one where changing framerates fixes the problem?

Not reproducible. Also, this is specific to the Direct3D 11 mode, it works fine on OpenGL.

**Comment by Layendan** [comment]
> Not reproducible. Also, this is specific to the Direct3D 11 mode, it works fine on OpenGL.

I have the same problem with OpenGL as well as the Legacy render.

**Comment by peppy** [comment]
There's a high likelihood that you have a driver setting overriding things.

**Comment by Layendan** [comment]
> There's a high likelihood that you have a driver setting overriding things.

Although I would agree with that, I never changed any of my driver settings.

I tried resetting them when it was mentioned in this thread, but that didn't change anything.

I've been playing this game for 2 years with no problem, then I boot up the game and the issue starts showing up.

**Comment by Kulei666** [comment]
Have the same issue, never changed any of my driver settings. It happened after an update, but I am not sure what version I was updating from. Regardless of the VSync/2x/4x/8x/unlimited setting FPS draw drops to 60/60, with ~17ms of lag input, both on keyboard and mouse. It seems to happen whenever I click play and it loads up maps (my fps drops to 1 and then it only bounces back to 60) and whenever I start a beatmap (same). Temporary fix is to pause the beatmap and alt+tab and the FPS bounces back to the set value, however if there is any stuttering during gameplay (even above the max fps limit, best visible with uncapped) the game might snap to 60 for some reason. It happens both in DirectX and OpenGL. Very rarely it uncaps itself during gameplay for a few seconds, but then usually snaps back to the mysterious 60.

**Comment by Layendan** [comment]
So I've been playing with a few apps that I have and am planning on doing a full wipe of my pc to test some more, but I've noticed that if I have a chromium based browser (might be any browser, but only tested with Chrome and Edge) open while opening the game, it will consistently cap the fps, while if the browser is closed it will only do it once (Tested only on Legacy OpenGL). Then you can alt+tab out and back in, then the fps does not cap anymore.

@Kulei666 what browser are you using and would you mind please trying this out to see if that works for you as well?

**Comment by peppy** [comment]
This sounds like you are

- Running borderless or windowed
- Have the chrome instance open on a second display with 60hz refresh rate.

This is a known issues on windows.

**Comment by Kulei666** [comment]
@Layendan @peppy Did some more testing today, turned off all apps and changed NVIDIA's control panel performance mode into Max Performance. None of that helped. Alt+tabbing still solves the issue (I am playing on fullscreen). Closely examining the fps and track length I made an observation (on two 6-star maps) that it capped 3 times, at around 5%, 40% and 85% of the map. and it usually caps after a beatmap break (by break i mean while music is playing but no circles), what happens during a break is that fps first drop and cap at 180, then drop and cap at 120 and when the beatmap resumes from the break it drops to 60 fps again. The maps I tested were: (but it of course happens on other maps too, I am not yet sure if its at around 5%, 40% and 85% of the map there too)
![image](https://github.com/ppy/osu/assets/123197787/298bc9fd-4560-423f-b755-585fcaaf5f6b)
![image](https://github.com/ppy/osu/assets/123197787/2e457c4a-0f95-428d-a6a5-579fc23d2465)

**Comment by peppy** [comment]
if it's exactly 69 then it's vsync

**Comment by Kulei666** [comment]
Yes, it's exactly 60 fps, it should be 8x that, 480fps. More testing and: When it drops below 420fps it immediatelly snaps to 60fps. 3 seconds before a break end it drop to 180, 1 seconds before break end it drops to 120, 2 seconds into the map it drops to 60fps and snaps perfectly onto it until alt+tabbed. How do I fix that? I forced Vertical sync Off in the Nvidia control panel and don't have Vsync active in osu!lazer

**Comment by Kulei666** [comment]
Another weird fix - recording the display with OBS makes the problem not happen at all, doesn't snap to 60fps.
